coderd/database/dbtestutil/driver.go

```diff
@@ -0,0 +1,127 @@
+package dbtestutil
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"fmt"
+
+	"github.com/lib/pq"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/cryptorand"
+)
+
+var (
+	_ driver.Driver             = &Driver{}
+	_ database.ConnectorCreator = &Driver{}
+	_ database.DialerConnector  = &Connector{}
+)
+
+type Driver struct {
+	name        string
+	inner       driver.Driver
+	connections []driver.Conn
+	listeners   map[chan struct{}]chan struct{}
```

Comment:
This connections and listeners business is not threadsafe, and so can cause races in our tests.

A better design would be to just have a `chan driver.Conn` that we pass each new connection thru.  Tests will have to know how many connections they expect, and can read from the channel. For the `PubSub`, it's the listener we want to test, which should use a connection when it first connects, then another one when it reconnects.

That way you don't need methods like `AddConnection`, `WaitForConnection`, or `DropConnections` -- the test code reads from the channel to wait for the connection, and can directly close the connection when it wants to interrupt it.

There is some complexity around the `sql.DB`, which we use for publishing and has a pool of connections. I suggest you sidestep that complexity by just using a second `PubSub` for publishing with a regular `pq` driver.